nomad/blocked_evals.go

```diff
@@ -0,0 +1,303 @@
+			// and also any eval that is not eligible or uneligible. This signifies that
+			// when the evaluation was originally run through the scheduler, that it
+			// never saw a node with the given computed class and thus needs to be
+			// unblocked for correctness.
+			var untrack []string
+			for id, eval := range b.captured {
+				if elig, ok := eval.ClassEligibility[computedClass]; ok {
+					if !elig {
+						// Can skip because the eval has explicitely marked the node class
+						// as ineligible.
+						continue
+					}
+				}
+
+				// The computed node class has never been seen by the eval so we unblock
+				// it.
+				unblocked = append(unblocked, eval)
+				untrack = append(untrack, id)
+				delete(b.jobs, eval.JobID)
+			}
+
+			// Untrack the unblocked evals.
+			if l := len(untrack); l != 0 {
+				for _, id := range untrack {
+					delete(b.captured, id)
+				}
+			}
+
+			if l := len(unblocked); l != 0 {
+				// Update the counters
+				b.stats.TotalEscaped = 0
+				b.stats.TotalBlocked -= l
+
+				// Enqueue all the unblocked evals into the broker.
+				b.evalBroker.EnqueueAll(unblocked)
+			}
+			b.l.Unlock()
+		}
+	}
+}
+
+// GetDuplicates returns all the duplicate evaluations and blocks until the
+// passed timeout.
+func (b *BlockedEvals) GetDuplicates(timeout time.Duration) []*structs.Evaluation {
+	var timeoutTimer *time.Timer
+	var timeoutCh <-chan time.Time
+SCAN:
+	b.l.Lock()
+	if len(b.duplicates) != 0 {
+		dups := b.duplicates
+		b.duplicates = nil
+		b.l.Unlock()
+		return dups
+	}
+	b.l.Unlock()
+
+	// Create the timer
+	if timeoutTimer == nil && timeout != 0 {
+		timeoutTimer = time.NewTimer(timeout)
+		timeoutCh = timeoutTimer.C
+		defer timeoutTimer.Stop()
+	}
+
+	select {
+	case <-b.stopCh:
+		return nil
+	case <-timeoutCh:
+		return nil
+	case <-b.duplicateCh:
+		goto SCAN
+	}
+
+	return nil
+}
+
+// Flush is used to clear the state of blocked evaluations.
+func (b *BlockedEvals) Flush() {
+	b.l.Lock()
+	defer b.l.Unlock()
+
+	// Kill any running goroutines
+	if b.running {
+		close(b.stopCh)
+		b.running = false
+	}
+
+	// Reset the blocked eval tracker.
+	b.stats.TotalEscaped = 0
+	b.stats.TotalBlocked = 0
+	b.captured = make(map[string]*structs.Evaluation)
+	b.escaped = make(map[string]*structs.Evaluation)
+	b.jobs = make(map[string]struct{})
+	b.duplicates = nil
+	b.unblockCh = make(chan string, unblockBuffer)
+	b.stopCh = make(chan struct{})
+	b.duplicateCh = make(chan struct{})
+}
+
+// Stats is used to query the state of the blocked eval tracker.
+func (b *BlockedEvals) Stats() *BlockedStats {
+	// Allocate a new stats struct
+	stats := new(BlockedStats)
+
+	b.l.RLock()
+	defer b.l.RUnlock()
+
+	// Copy all the stats
+	stats.TotalEscaped = b.stats.TotalEscaped
+	stats.TotalBlocked = b.stats.TotalBlocked
+	return stats
+}
+
+// EmitStats is used to export metrics about the blocked eval tracker while enabled
+func (b *BlockedEvals) EmitStats(period time.Duration, stopCh chan struct{}) {
```

Comment:
Good catch, forgot to add the call. You want it to be the servers shutdown channel because otherwise you'll stop emitting stats when your not the leader and it won't start again even if you become a leader again.